src/category_theory/sites/dense_subsite.lean

```diff
@@ -0,0 +1,412 @@
+not well-behaved otherwise. Note that https://ncatlab.org/nlab/show/dense+sub-site indeed has a
+weaker notion of cover-dense that loosen this requirement, but it would not have all the properties
+we would need, and some sheafification would be needed for here and there.
+
+## Main results
+
+- `category_theory.cover_dense.sheaf_hom`: If `G : C ⥤ (D, K)` is full and cover-dense,
+  then given any presheaf `ℱ` and sheaf `ℱ'` on `D`, and a morphism `α : G ⋙ ℱ ⟶ G ⋙ ℱ'`,
+  we may glue them together to obtain a morphism of sheaves `ℱ ⟶ ℱ'`.
+- `category_theory.cover_dense.sheaf_iso`: If the `α` above is iso, then the result is also iso.
+- `category_theory.cover_dense.iso_of_restrict_iso`: If `G : C ⥤ (D, K)` is full and cover-dense,
+  then given any sheaves `ℱ, ℱ'` on `D`, and a morphism `α : ℱ ⟶ ℱ'`, then `α` is an iso if
+  `G ⋙ ℱ ⟶ G ⋙ ℱ'` is iso.
+
+## References
+
+* [Elephant]: *Sketches of an Elephant*, ℱ. T. Johnstone: C2.2.
+* https://ncatlab.org/nlab/show/dense+sub-site
+* https://ncatlab.org/nlab/show/comparison+lemma
+
+-/
+
+universes v
+
+namespace category_theory
+
+variables {C : Type*} [category C] {D : Type*} [category D] {E : Type*} [category E]
+variables (J : grothendieck_topology C) (K : grothendieck_topology D)
+variables {L : grothendieck_topology E}
+
+/--
+An auxiliary structure that witnesses the fact that `f` factors through an image object of `G`.
+-/
+@[nolint has_inhabited_instance]
+structure presieve.cover_by_image_structure (G : C ⥤ D) {V U : D} (f : V ⟶ U) :=
+(obj : C) (lift : V ⟶ G.obj obj) (map : G.obj obj ⟶ U) (fac : lift ≫ map = f)
+
+attribute [simp, reassoc] presieve.cover_by_image_structure.fac
+
+/--
+Given a functor `G`, the presieve of `U : D` such that each arrows factors through images of `G`.
+-/
+def presieve.cover_by_image (G : C ⥤ D) (U : D) : presieve U :=
+λ Y f, nonempty (presieve.cover_by_image_structure G f)
+
+/--
+Given a functor `G`, the sieve of `U : D` such that each arrow factors through images of `G`.
+-/
+def sieve.cover_by_image (G : C ⥤ D) (U : D) : sieve U :=
+⟨presieve.cover_by_image G U,
+  λ X Y f ⟨⟨Z, f₁, f₂, eq⟩⟩ g, ⟨⟨Z, g ≫ f₁, f₂, by rw [category.assoc, ← eq] ⟩⟩⟩
+
+lemma presieve.in_cover_by_image (G : C ⥤ D) {X : D} {Y : C} (f : G.obj Y ⟶ X) :
+  presieve.cover_by_image G X f := ⟨⟨Y, 𝟙 _, f, by simp⟩⟩
+
+/--
+A functor `G : (C, J) ⥤ (D, K)` is called `cover_dense` if for each object in `D`,
+  there exists a covering sieve in `D` that factors through images of `G`.
+
+This definition can be found in https://ncatlab.org/nlab/show/dense+sub-site Definition 2.2.
+-/
+structure cover_dense (K : grothendieck_topology D) (G : C ⥤ D) : Prop :=
+(is_cover : ∀ (U : D), sieve.cover_by_image G U ∈ K U)
+
+open presieve opposite
+namespace cover_dense
+variables {A : Type*} [category A] {K} {G : C ⥤ D} (H : cover_dense K G)
+
+lemma ext (H : cover_dense K G) (ℱ : SheafOfTypes K) (X : D) {s t : ℱ.val.obj (op X)}
```

Comment:
Do you want the `ext` tactic to automatically pick up this lemma? If so, please tag it with `@[ext]`.

Reply:
This lemma needs a `cover_dense` instance at hand that could not be inferred from the result type, so I think it is not possible to do so.

Reply:
Good point.
```suggestion
-- this is not marked with `@[ext]` because `H` can not be inferred from the type
lemma ext (H : cover_dense K G) (ℱ : SheafOfTypes K) (X : D) {s t : ℱ.val.obj (op X)}
```